### ISSUE TITLE
Crash in Fl_Preferences if file has bad format

### ISSUE THREAD

**Issue by MatthiasWM** [fltk/fltk#891]
https://groups.google.com/g/fltkcoredev/c/ZyP9WGrTtw8/m/KfiCua-hAgAJ

**Comment by MatthiasWM** [comment]
Andre: "just run into a segv in the Fl_Preferences::RootNode::read method due to null pointer dereference in line 1333 :
       nd->set( buf );
it happened with the latest snapshot fltk-1.4.x-20240112-421e917f probably with an invalid ~/.fltk/[fltk.org/filechooser.prefs](http://fltk.org/filechooser.prefs) nevertheless fltk should not just crash. never changed filechooser.prefs manually, only via the fltk file_chooser dialog. it somehow happened due to apps using different versions of fltk 1.3.x and 1.4.x attached patch adds checks for null value of nd in Fl_Preferences::RootNode::read"
```
diff -Nur fltk-1.4.x-20240112-421e917f/src/Fl_Preferences.cxx fltk-1.4.x-20240112/src/Fl_Preferences.cxx
--- fltk-1.4.x-20240112-421e917f/src/Fl_Preferences.cxx	2024-01-11 17:21:49.000000000 +0100
+++ fltk-1.4.x-20240112/src/Fl_Preferences.cxx	2024-01-14 17:00:04.000000000 +0100
@@ -1324,13 +1324,13 @@
       size_t end = strcspn( buf+1, "\n\r" );
       if ( end != 0 ) {                         // if entry is not empty
         buf[ end+1 ] = 0;
-        nd->add( buf+1 );
+        if(NULL != nd) { nd->add( buf+1 ); }
       }
     } else {                                     // read a name/value pair
       size_t end = strcspn( buf, "\n\r" );
       if ( end != 0 ) {                         // if entry is not empty
         buf[ end ] = 0;
-        nd->set( buf );
+        if(NULL != nd) { nd->set( buf ); }
       }
     }
   }
```

**Comment by Albrecht-S** [comment]
> it somehow happened due to apps using different versions of fltk 1.3.x and 1.4.x

@MatthiasWM I noticed such effects myself, hence this is related:

I suggest (and request) to invert the search order for existing `Fl_Preferences` files under Unix/Linux regarding the "old" storage location and the "new" one based on `XDG*` variables for better coexistence of FLTK 1.3 and FLTK 1.4 apps on the same (Linux) system. This could be relevant for instance for Debian which tends to install old and new versions in parallel. There's no similar option for other platforms like Windows and macOS in my proposal though.

My proposal is:
- FLTK 1.4 should **_prefer_** [1] the new XDG based location
- FLTK 1.3 searches (anyway) always in the old location.

[1] "**_Prefer_**" means in this context to search for `Fl_Preferences` files **first** in the new location and use this location for all preferences if `~/.config/fltk.org/` (or any other XDG based directory named `fltk.org`) exists. Currently FLTK 1.4 searches first in the old location.

**_Details and Rationale:_**

The "old" (1.3.x) location is under `~/.fltk/…` whereas the "new" (XDG based) location is typically under `~/.config/fltk.org/`. The documentation recommends to **move** the FLTK preferences files from the 1.3.x location to the XDG based location for usage with FLTK 1.4 (at least this was the reason for the new location based on XDG*). This works fine, and I did this on my own system. Silent upgrades to new variables, formats, and features (e.g. Fluid's _scheme name_ rather than _numeric scheme index_) work fine when using the new location.

However, as soon as you execute an old 1.3 application (which I did of course for testing) Fl_Preferences files are created under the "old" (1.3) location. As it stands now, FLTK 1.4 applications will subsequently "find" the preferences files in the old location (maybe in 1.3 "format" rather than 1.4 "format") and both applications are overwriting each other's preferences files.

This may have led to the effect the OP described and could be mitigated by only changing the search order. I don't have a suggestion for other platforms though unless we use "versioned" Fl_Preferences locations (directory names) in general. The problem is that we can't assume that moving from 1.3 to 1.4 is done at one point in time and once forever. There may be coexistence of 1.3 and 1.4 applications on the same system.

I would fix the search order myself but I don't think that I can afford the time right now before the release. Maybe you can take a look? I did and noticed that the code is pretty "complicated" because it has to handle several directories and files but since you wrote it initially... Thanks in advance.

**Comment by MatthiasWM** [comment]
I can change the order, no problem. Not sure if it solves the original problem. 

The format did not change between 1.3 and 1.4 to my knowledge.

So I assume if no file is found at the 1.4 location, but at the 1.3 location, I obviously read from 1.3. Should I *write* to 1.3 as well, or create a new file at the 1.4 location?

**Comment by Albrecht-S** [comment]
> I can change the order, no problem.

Yes, please.

> Not sure if it solves the original problem.

It won't "solve" the original problem, it may only prevent such issues in the future. But we should avoid - now and in the future - that 1.3 apps and 1.4 apps write to the same preferences files [1] (Debian has still FLTK 1.1 and 1.3 and both can be installed in parallel IIRC).

> The format did not change between 1.3 and 1.4 to my knowledge.

Not the format itself but maybe the contents. And what do we know about what we'll add or change in the future?

As you may have noticed (or not?) I changed the "scheme" stored in fluid's prefs from the _scheme index_ (used in 1.3) to the _scheme name_ in 1.4 (1.4 reads the index only if the name doesn't exist and stores both the name and the index for backwards compatibility). Now we'd get inconsistent data if an old (FLTK 1.3) fluid was used and stored another scheme **index** later [1]. This is only one example.

> So I assume if no file is found at the 1.4 location, but at the 1.3 location, I obviously read from 1.3. Should I write to 1.3 as well, or create a new file at the 1.4 location?

No, we should always write back to the file we read. It should be the user's choice to **move** their preferences files to the new location. And with "the new location" I mean the top level FLTK Fl_Preferences directory. It would be possible that a user moves all existing preferences files to the new location but later they run e.g. `blocks` in the FLTK 1.3 version. This would create `blocks.prefs` in the "old location". Now this is the only prefs file in the old location and all FLTK 1.4 apps won't see their own preferences anymore (because they try to open all prefs from the old location just because it exists).

Note that I don't want to check the location per file (application), only the top level prefs folder. The former would really contribute to confusion.

-----

[1] That said, this "pragmatic solution" (or, say: proposal) was born because I encountered such issues on my Linux system. It would be fine if my FLTK 1.4 tests would all read/write from/to the new location and all 1.3 apps from the old location. However, this is not optimal because this would only work on Linux and only if the user moved their preferences files from the old to the new location. Similar things could be done on Windows where we also changed the location.
**_A better (proper) solution would be to use "versioned" storage locations_** so we don't need to take care of 1.1/1.3 formats vs 1.4 vs 1.5 formats.

**Comment by MatthiasWM** [comment]
> > So I assume if no file is found at the 1.4 location, but at the 1.3 location, I obviously read from 1.3. Should I write to 1.3 as well, or create a new file at the 1.4 location?
> 
> No, we should always write back to the file we read. 

Ok, thanks.

> Note that I don't want to check the location per file (application), only the top level prefs folder. The former would really contribute to confusion.

For example, `blocks` calls `Fl_Preferences(Fl_Preferences::USER_L, "fltk.org", "blocks")`. So I would check for the existence of 
`$XDG_CONFIG_HOME/fltk.org/` or `~/.config/fltk.org/` if `XDG_CONFIG_HOME` is not set. If the directory exists, we create and/or read `blocks.prefs` there.

If not, I check for `~/.fltk/fltk.org/`. If that directory exists, I read/write `~/.fltk/fltk.org/blocks.prefs`.

And if neither is the case, I create `$XDG_CONFIG_HOME/fltk.org/blocks.prefs`. Correct?
 
> Similar things could be done on Windows where we also changed the location.

That is no problem, if you think it is better that way. I'll do that too then.

> **_A better (proper) solution would be to use "versioned" storage locations_** so we don't need to take care of 1.1/1.3 formats vs 1.4 vs 1.5 formats.

Not sure if I get it. Do you want me to add a flag that forces the use of the 1.3 path?

**Comment by Albrecht-S** [comment]
> For example, blocks calls Fl_Preferences(Fl_Preferences::USER_L, "fltk.org", "blocks"). So I would check for the existence of $XDG_CONFIG_HOME/fltk.org/ or ~/.config/fltk.org/ if XDG_CONFIG_HOME is not set. If the directory exists, we create and/or read blocks.prefs there.
> If not, I check for ~/.fltk/fltk.org/. If that directory exists, I read/write ~/.fltk/fltk.org/blocks.prefs.
> And if neither is the case, I create $XDG_CONFIG_HOME/fltk.org/blocks.prefs. Correct?

Yes, that sounds good. With the exception that there's also a default value if `XDG_CONFIG_HOME` is not set. In this case we'd also check `$HOME/.config/fltk.org` directly and use this if it exists. If neither exists we'd fall back to the 1.3 location. If that doesn't exist we should use a pristine `$XDG_CONFIG_HOME` or `$HOME/.config/fltk.org` top directory and create everything there.

I believe it's "as simple as" (TM):

1. Does the XDG based top level folder 'fltk.org' exist? If yes, use it.
2. If not: does the old location `$HOME/.fltk/fltk.org` exist? If yes, use it.
3. If not: fall back to (1.) and use the XDG based folder (create it and the prefs file below it).

After that FLTK 1.4 apps will always use the XDG based prefs location, even if an FLTK 1.3 app creates the old `$HOME/.fltk/fltk.org` later. This will have the side effect that `FLTK 1.4 blocks` will have a different max score than `FLTK 1.3 blocks` and `fluid 1.4` can use a more modern scheme ('oxy') than `fluid 1.3` that doesn't even know the `oxy` scheme. And so on.

>> A better (proper) solution would be to use "versioned" storage locations ...

>  Do you want me to add a flag that forces the use of the 1.3 path?

No. Sorry, this becomes now more and more OT here, maybe I shouldn't have started this discussion.

I'm thinking further: what do we do after `$XDG_CONFIG_HOME/fltk.org` has been established and stabilized, or the new location on Windows, and whatever on macOS. We may have FLTK 1.3, 1.4, and also 1.5 or later coexisting on one system. Developers may use `CMake/FetchContent` to pull in **_any version of FLTK_** and build their own applications. A common thing to all are the `fltk-options` and their new program. What if we change option values (semantics), add new ones, etc. ? If there was only one `.prefs` file for all FLTK versions, then programs would overwrite each others' preferences files with potentially incompatible contents.

A "versioned" location could be derived by using e.g. `$XDG_CONFIG_HOME/fltk.org/1.3/` or `$XDG_CONFIG_HOME/fltk.org/1.4/` or something like this. CMake's Modules directories also include the CMake version (e.g. `/usr/share/cmake-3.25/Modules/Find*.cmake`), that's what I thought about.

However, now may not be the right time and this issue is likely not the right place to discuss this further. We can check this later elsewhere. [Versioned library and include file locations are another issue, BTW, but not for now, sorry.]

**Comment by MatthiasWM** [comment]
The path search order is now fixed for 1.4 . The MSWindows path is computed differently, but it is still the same path, so I did not change that.

**Comment by Albrecht-S** [comment]
I tested the Unix path search and found an issue with a wrong path.

Fixed in commit 667adac6ddbbf26c8098321fe017e1e0d4b0f3a9.
With this commit I believe it's fixed now. Can we close?

**Comment by wcout** [comment]
I notice that with the last commits for this issue the preferences I already had for FLTK 1.4 (e.g. favorites in filechooser) have disappared. Is this intentional?

**Comment by Albrecht-S** [comment]
> I notice that with the last commits for this issue the preferences I already had for FLTK 1.4 (e.g. favorites in filechooser) have disappared. Is this intentional?

Certainly not intentional.

There have been some problems with the first two commits regarding this issue. I believe that I fixed the last issue in commit 5cc5a60e096f96924e1409bb410a978db025876a.

@wcout Did you "move" your preferences files from `~/.fltk` (_old location_) to `$XDG_CONFIG_HOME/fltk.org` or its default folder `~/.config/fltk.org` (_new location_)? If you did **and** used both FLTK 1.3 and FLTK 1.4 applications after the move, then the following scenario may have happened:

1. After the move FLTK 1.4 used the _new location_ as intended.
2. When you used the first FLTK 1.3 program this program created its preferences in the _old location_.
3. After that all FLTK 1.4 used the _old location_ again, "**forgetting**" all prefs in the _new location_, and FLTK 1.3 and 1.4 apps would overwrite each others' preferences files. This could have unforeseen effects if new preferences would be added in FLTK 1.4 etc..
4. You might have changed your file chooser preferences at this time (with either FLTK 1.3 or 1.4).
5. With the latest commits FLTK 1.4 uses the _new location_ (again) even though preferences in the _old location_ exist and may have been changed.

The intention of this fix is to avoid point 3 caused by point 2, i.e. once you moved your preferences to the _new location_ FLTK 1.4 will **exclusively use the _new location_** and FLTK 1.3 will use the old location as before. This may appear as a loss of preferences in FLTK 1.3 applications. As of now it would sometimes be better to **copy** the preferences from the _old location_ to the _new location_ because then FLTK 1.3 and 1.4 would **initially see the same preferences** - although they would **store changes in different locations**.

Between commits baa3c6ddc0cf8136ede5fb08acec54675cdc5582 and 5cc5a60e096f96924e1409bb410a978db025876a some unintended (buggy) behavior may have happened [1]. You may want to check which preferences files you have in the _old_ and _new locations_, and if you do, please keep in mind that only the top-level folder is used to decide which preferences tree is to be used. If both exist, then FLTK 1.3 and FLTK 1.4 applications use **_distinct preferences_** for "everything".

-----

[1] If you had moved your preferences to the new location and the old location was empty it might have happened that FLTK 1.4 apps used between these two commits may have created the old preferences folder again but I'm not sure. Something like this did happen on my system, anyway, but I tested it and fixed the bug (I created and deleted old and new prefs folders while testing, hence I can't tell exactly what had happened or could have happened on other systems).

If you **_did not move your prefs_** from the _old location_ to _new location_: ISTR that the fallback to the old location didn't work at that time. Hence FLTK 1.4 apps may have created their prefs in the _new location_ although prefs existed in the _old location_ (which was not intended). Now FLTK 1.4 uses the new location and does no longer see your old prefs. If this happened to you the best way out would be to delete the entire new folder (`~/.config/fltk.org`) and this should resurrect your old preferences.

**Comment by Albrecht-S** [comment]
Hint: I like to use `test/preferences(.cxx)` to test which preferences folder is used.
- The 1.4 version outputs the folder on stdout so you can directly see it.
- The 1.3 version doesn't output the folder but you can easily change some prefs and see what happens.

**Comment by Albrecht-S** [comment]
> Is that correct?

Yep. As long as there is only the old location `~/.fltk/fltk.org` it should be kept and used. The first change was requested by a user to honor the XDG based config path `~/.config/fltk.org` and this is use only if the user actively moves (or copies) the preferences files to the new location. Or, if there are _no prefs at the old location_ **and** _the latest FLTK 1.4 executes and creates new prefs_. In your case the new prefs were obviously created by accident and were shadowing the prefs in the old location.
